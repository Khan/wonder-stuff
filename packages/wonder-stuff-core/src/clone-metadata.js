// @flow
import type {Metadata} from "./types.js";

const cloneValue = <T>(value: T): T => {
    if (value != null && typeof value === "object") {
        if (Array.isArray(value)) {
            return value.map(cloneValue);
        } else {
            return Object.keys(value).reduce((acc, key) => {
                acc[key] = cloneValue(value[key]);
                return acc;
                // We need to cast to any here so that the value is a valid
                // match for T in the return type after the accumulator is done.
                // $FlowIgnore[unclear-type]
            }, ({}: any));
        }
    }
    return value;
};

/**
 * Clone the given metadata.
 *
 * This performs a deep clone of the given metadata and returns a read-only
 * clone.
 */
export const cloneMetadata = <T: Metadata>(
    metadata: ?$ReadOnly<T>,
): ?$ReadOnly<T> => {
    // If it's null or undefined, just return it as-is.
    if (metadata == null) {
        return metadata;
    }

    // Clone the data.
    const clone = cloneValue(metadata);

    // We know that we have cloned the incoming data and so our clone is of
    // the correct type.
    // $FlowIgnore[incompatible-return]
    return Object.freeze(clone);
};
