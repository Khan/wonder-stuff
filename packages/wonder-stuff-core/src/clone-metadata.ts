import {clone} from "./clone";
import type {Metadata} from "./types";

/**
 * Clone the given metadata.
 *
 * This performs a deep clone of the given metadata and returns a read-only
 * clone.
 */
export const cloneMetadata = <T extends Metadata>(metadata?: Readonly<T> | null): Readonly<T> | null | undefined => {
    // If it's null or undefined, just return it as-is.
    if (metadata == null) {
        return metadata;
    }

    // Clone the data.
    const clonedValue = clone(metadata);

    // We know that we have cloned the incoming data and so our clone is of
    // the correct type.
    return Object.freeze(clonedValue);
};
