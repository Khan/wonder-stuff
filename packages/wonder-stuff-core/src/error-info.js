// @flow
import {buildCausedByMessage} from "./build-caused-by-message.js";

/**
 * Information about an error.
 *
 * This is used to simplify and standardize extracting and combining error
 * information.
 */
export class ErrorInfo {
    +_name: string;
    +_message: string;
    +_stackFrames: $ReadOnlyArray<string>;

    constructor(
        name: string,
        message: string,
        stackFrames: $ReadOnlyArray<string>,
    ) {
        this._name = name;
        this._message = message;
        // Copy the stack array just to be sure our internal state won't
        // mutate outside our control.
        this._stackFrames = [...stackFrames];
    }

    /**
     * The error message.
     */
    get message(): string {
        return this._message;
    }

    /**
     * The error name.
     */
    get name(): string {
        return this._name;
    }

    /**
     * The error message prefixed with the error name.
     */
    get messageWithName(): string {
        return `${this.name}: ${this.message}`;
    }

    /**
     * An array representing the stack frames of the error.
     *
     * Index 0 being the top of the stack.
     */
    get stack(): $ReadOnlyArray<string> {
        // Copy the stack array just to be sure our internal state won't
        // mutate outside our control.
        return [...this._stackFrames];
    }

    /**
     * Get the stack trace including the error name and message.
     *
     * This includes the error name and message at the start as is standard
     * in many platforms, thus standardizing how we represent this info.
     */
    get standardizedStack(): string {
        // Now construct the string value.
        return `${this.messageWithName}\n${this._stackFrames.join("\n")}`;
    }

    /**
     * Combine the info of an error and the error that caused it.
     *
     * This combines the stacks of a consequential error and the error that
     * caused it, creating a new stack trace that incorporates the frames of
     * both errors to tell a deeper story of what happened.
     *
     * For example, some work is done and error A gets thrown, this is caught
     * and ultimately wrapped by error B, which then gets thrown. By combining
     * the stack traces of both error A and B we can get a better idea of what
     * happened, which can be especially useful in asynchronous code.
     *
     * Stack traces are combined such that common frames are deduped, and then
     * at the point of the original throw of error A (i.e. error A's top stack
     * frame), the deduped trace of error B is inserted to show what happened
     * after the error was caught.
     *
     * So, given:
     *
     * cause = ErrorA, with stack,
     *    Error: A
     *      stackA2
     *      stackA1
     *      stackShared2
     *      stackShared1
     *
     * and consequence = ErrorB, with stack,
     *    Error: B
     *       stackB2
     *       stackB1
     *       stackShared2
     *       stackShared1
     *
     * This method would produce the equivalent of:
     *    Error: B
     *        caused by
     *            Error: A
     *      stackA2
     *      stackA1
     *      stackB2
     *      stackB1
     *      stackShared2
     *      stackShared1
     */
    static fromConsequenceAndCause(
        consequence: ErrorInfo,
        cause: ErrorInfo,
    ): ErrorInfo {
        if (process.env.NODE_ENV !== "production") {
            // Verify our arguments.
            if (!(consequence instanceof ErrorInfo)) {
                throw new Error("consequence must be an instance of ErrorInfo");
            }
            if (!(cause instanceof ErrorInfo)) {
                throw new Error("cause must be an instance of ErrorInfo");
            }
            if (cause === consequence) {
                throw new Error("cause and consequence must be different");
            }
        }

        // Now, let's combine the stacks. First, we dedupe the frames they
        // share.
        const combinedStackFrames = [];
        const stackFramesA = cause.stack;
        const stackFramesB = consequence?.stack ?? [];

        // Shared stack frames will be at the bottom of each stack of frames.
        let indexA = stackFramesA.length - 1;
        let indexB = stackFramesB.length - 1;

        // Now iterate both stacks, tracking the frames that exist in both.
        while (
            indexA >= 0 &&
            indexB >= 0 &&
            stackFramesA[indexA] === stackFramesB[indexB]
        ) {
            combinedStackFrames.unshift(stackFramesA[indexA]);
            indexA--;
            indexB--;
        }

        // At this point, either we've exhausted one or both stacks, or we need
        // to add frames from A and then remaining frames from B.
        for (indexA; indexA >= 0; indexA--) {
            combinedStackFrames.unshift(stackFramesA[indexA]);
        }
        for (indexB; indexB >= 0; indexB--) {
            combinedStackFrames.unshift(stackFramesB[indexB]);
        }

        // Now that we've combined the stacks, let's also combine the messages
        // and return our new stack.
        return new ErrorInfo(
            consequence.name,
            buildCausedByMessage(consequence.message, cause.messageWithName),
            combinedStackFrames,
        );
    }

    /**
     * Get normalized error information from an error.
     *
     * @param {Error} error The error to be normalized.
     * @param {number} [stripFrames] The number of stack frames to drop from
     * the top of the stack. Defaults to `0`.
     * @param {number} [minimumFrameCount] The number of stack frames to keep
     * as a minimum. This overrides the `stripFrames` value and defaults to `1`.
     * @returns {ErrorInfo} Normalized error information for the given
     * error.
     */
    static normalize(
        error: Error,
        stripFrames: number = 0,
        minimumFrameCount: number = 1,
    ): ErrorInfo {
        if (process.env.NODE_ENV !== "production") {
            // Verify our arguments.
            if (!(error instanceof Error)) {
                throw new Error("Error must be an instance of Error");
            }
            if (stripFrames < 0) {
                throw new Error("stripFrames must be >= 0");
            }
            if (minimumFrameCount < 0) {
                throw new Error("minimumFrameCount must be >= 0");
            }
        }

        const errorMessage = error.message;

        // The normalized message is just the first line of the error message.
        const normalizedMessage =
            errorMessage
                .toString()
                .split("\n")
                .find((l) => l.trim().length) ?? "(empty message)";

        // OK, get the stack without the error name and message
        // (unless they are the same).
        const fullErrorMessage = error.toString();
        const stackWithoutMessage =
            error.stack?.startsWith(fullErrorMessage) &&
            error.stack !== fullErrorMessage
                ? error.stack.substring(fullErrorMessage.length)
                : error.stack ?? "";

        // Now split those frames into individual frame lines, filtering
        // out any lines that are solely whitespace.
        const stackFrames = stackWithoutMessage
            .split("\n")
            .filter((s) => s.trim().length);

        // Calculate the real number of frames we strip, taking into account
        // the minimum number of frames we want.
        const actualStripFrames =
            stackFrames.length >= stripFrames + minimumFrameCount
                ? stripFrames
                : 0;

        return new ErrorInfo(
            error.name,
            normalizedMessage,
            stackFrames.slice(actualStripFrames),
        );
    }

    /**
     * Get error information from an error.
     *
     * @param {Error} error The error from which to grab the info.
     * @returns {ErrorInfo} Error information for the given error.
     */
    static from(error: Error): ErrorInfo {
        if (process.env.NODE_ENV !== "production") {
            // Verify our arguments.
            if (!(error instanceof Error)) {
                throw new Error("Error must be an instance of Error");
            }
        }

        // OK, get the stack without the error message (unless they are the same).
        const fullErrorMessage = error.toString();
        const stackWithoutMessage =
            error.stack?.startsWith(fullErrorMessage) &&
            error.stack !== fullErrorMessage
                ? error.stack.substring(fullErrorMessage.length)
                : error.stack ?? "";

        // Now split those frames into individual frame lines, filtering
        // out any lines that are solely whitespace.
        const stackFrames = stackWithoutMessage
            .split("\n")
            .filter((s) => s.trim().length);

        return new ErrorInfo(error.name, error.message, stackFrames);
    }
}
