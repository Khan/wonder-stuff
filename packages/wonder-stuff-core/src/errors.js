// @flow

/**
 * @typedef {Object} Errors A base error taxonomy for consumers of this core
 * library to build off.
 *
 * @property {string} Unknown The kind of error is not known.
 */

/**
 * @type {Errors} A base error taxonomy.
 */
export const Errors = Object.freeze({
    Unknown: "Unknown",
});
