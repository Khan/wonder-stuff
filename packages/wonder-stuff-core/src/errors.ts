/**
 * @typedef {Object} Errors A base error taxonomy for consumers of this core
 * library to build off.
 *
 * @property {string} Unknown The kind of error is not known.
 * @property {string} Internal The error is internal to the executing code.
 * @property {string} InvalidInput There was a problem with the provided input,
 * such as the wrong format or a null value.
 * @property {string} InvalidUse The error is down an improper use of the
 * invoked code.
 */

/**
 * @type {Errors} A base error taxonomy.
 */
export const Errors = Object.freeze({
    Unknown: "Unknown",
    Internal: "Internal",
    InvalidInput: "InvalidInput",
    InvalidUse: "InvalidUse",
});
