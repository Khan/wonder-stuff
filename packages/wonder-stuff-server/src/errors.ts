import {Errors as CoreErrors} from "@khanacademy/wonder-stuff-core";

/**
 * @typedef {Object} Errors The different errors in our error taxononomy.
 *
 * @property {string} Unknown The kind of error is not known.
 *
 * @property {string} Internal The error is internal to the executing code.
 *
 * @property {ErrorKind} InvalidInput There was a problem with provided input,
 * such as the wrong format or a null value. This indicates inputs that are
 * problematic regardless of the state of the system. If the input was valid
 * but caused problems anyway consider {@link NotAllowed}.
 *
 * @property {string} InvalidUse The error is down an improper use of the
 * invoked code.
 *
 * @property {ErrorKind} NotFound A resource was not found. If the resource
 * couldn't be retrieved due to access control use {@link Unauthorized}
 * instead. If the resource couldn't be found because the input was invalid use
 * {@link InvalidInput} instead.
 *
 *
 * @property {ErrorKind} NotAllowed There was a problem due to the state of the
 * system not matching the requested operation or input. For example, trying to
 * create a username that is valid, but is already taken by another user. Use
 * {@link InvalidInput} instead when the input isn't valid regardless of
 * the state of the system. Use {@link NotFound} when the failure is due
 * to not being able to find a resource.
 *
 * @property {ErrorKind} Unauthorized There was an access control problem
 *
 * @property {ErrorKind} NotImplemented The function isn't implemented
 *
 * @property {ErrorKind} TransientService There was a problem when making a
 * request to a non-Khan service, e.g. datastore that might be resolvable by
 * retrying.
 *
 * @property {ErrorKind} Service There was a non-transient problem when making
 * a request to a non-Khan service, e.g. datastore.
 */

/**
 * @type {Errors} A base server error taxonomy.
 */
export const Errors = Object.freeze({
    ...CoreErrors,
    TransientService: "TransientService",
    Service: "Service",
});
