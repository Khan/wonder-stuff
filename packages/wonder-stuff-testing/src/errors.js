// @flow
import {Errors as CoreErrors} from "@khanacademy/wonder-stuff-core";

export const Errors = Object.freeze({
    ...CoreErrors,
    InvalidUse: "InvalidUse",
});
