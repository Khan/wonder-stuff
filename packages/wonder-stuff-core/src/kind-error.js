// @flow
import {Errors} from "./errors.js";
import {ErrorInfo} from "./error-info.js";
import {cloneMetadata} from "./clone-metadata.js";

import type {Metadata} from "./types.js";

/**
 * Options for constructing a `KindError`.
 */
type Options = {|
    /**
     * An error responsible for the error being created.
     *
     * @type {?Error}
     */
    cause?: ?Error,

    /**
     * Data to be attached to the error.
     *
     * @type {?$ReadOnly<Metadata>}
     */
    metadata?: ?$ReadOnly<Metadata>,

    /**
     * A prefix for the error name.
     *
     * @type {?string}
     */
    prefix?: ?string,

    /**
     * A name for the error.
     *
     * @type {?string}
     */
    name?: ?string,

    /**
     * The number of stack frames to strip from the error.
     *
     * @type {?number}
     */
    stripStackFrames?: ?number,

    /**
     * The minimum number of stack frames to try and retain.
     *
     * @type {?number}
     */
    minimumFrameCount?: ?number,
|};

/**
 * An error to describe detailed states and relationships.
 *
 * This error type supports error taxonomies and attachment of metadata, as
 * well as causal relationships between itself and another source error.
 * More specific variations can be built off this to provide a payload for
 * logging setups like Sentry or winston.
 */
export class KindError extends Error {
    +kind: string;
    +originalMessage: string;
    +metadata: ?$ReadOnly<Metadata>;
    +cause: ?Error;

    /**
     * Creates an instance of `KindError`.
     *
     * @memberof KindError
     * @param {string} message The error message.
     * @param {string} [kind] The kind of error. This will be combined with
     * `prefix` to form the name of the error, i.e. PrefixKindError.
     * Defaults to `Errors.Unknown`.
     * @param {Options} [options] Options for constructing the error.
     * @param {Error} [options.cause] The error that caused this error.
     * @param {$ReadOnly<Metadata>} [options.metadata] The metadata to attach
     * to the error.
     * @param {string} [options.prefix=""] A prefix to prepend the name of the
     * error.
     * @param {string} [options.name="Error"] The name of the error.
     * @param {number} [options.stripStackFrames=0] The number of stack frames
     * to remove from the error's stack. This can be used to ensure that the top
     * call of the stack references the point at which an error is thrown which
     * can be useful when helper functions are used to build the error being
     * thrown.
     * @param {number} [options.minimumFrameCount=1] The minimum number of
     * stack frames to try and retain. This can be used to prevent stripping
     * all stack frames from the error's stack.
     */
    constructor(
        message: string,
        kind: string = Errors.Unknown,
        {
            cause,
            prefix,
            name,
            metadata,
            stripStackFrames,
            minimumFrameCount,
        }: Options = {},
    ) {
        // Validate arguments.
        if (cause && !(cause instanceof Error)) {
            throw new Error("cause must be an instance of Error");
        }
        if (name != null && /\s/g.test(name)) {
            throw new Error("name must not contain whitespace");
        }
        if (/\s/g.test(kind)) {
            throw new Error("kind must not contain whitespace");
        }
        if (prefix != null && /\s/g.test(prefix)) {
            throw new Error("prefix must not contain whitespace");
        }
        if (stripStackFrames != null && stripStackFrames < 0) {
            throw new Error("stripStackFrames must be >= 0");
        }
        if (minimumFrameCount != null && minimumFrameCount < 0) {
            throw new Error("minimumFrameCount must be >= 0");
        }

        super(message);

        // Save the original message as we may change this message later.
        this.originalMessage = message;

        // Metadata associated with this error.
        this.metadata = cloneMetadata(metadata);

        // Set the name so we get a nice error output, like
        // KAInternalError
        this.name = `${prefix ?? ""}${kind}${name ?? ""}Error`;

        // The kind of error which can be used for grouping with
        // other error sources that use the same error taxonomy.
        this.kind = kind;

        // The cause of this error, if there is one.
        this.cause = cause;
        if (cause != null) {
            // We want to generate a better stack trace using the source error
            // and our own stack.
            const normalizedError = ErrorInfo.normalize(
                this,
                stripStackFrames ?? 0,
                minimumFrameCount ?? 1,
            );
            // We don't normalize the cause as this should have happened already
            // when it was, itself, created with a cause.
            const normalizedCause = ErrorInfo.from(cause);
            const combined = ErrorInfo.fromConsequenceAndCause(
                normalizedError,
                normalizedCause,
            );

            /**
             * $FlowIgnore[incompatible-type]
             * Flow doesn't like us leaving KAError temporarily without a
             * stack, but I don't want this crashing because we try to
             * assign to a readonly thing, so being mildly cautious.
             */
            delete this.stack;
            this.stack = combined.toString();

            // We update our message to a normalized one from the combined
            // stacks, as this gives us a nice causal relationship.
            this.message = combined.message;
        }
    }
}
