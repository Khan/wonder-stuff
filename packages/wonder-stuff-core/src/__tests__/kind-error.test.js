// @flow
import * as CloneMetadata from "../clone-metadata.js";
import {ErrorInfo} from "../error-info.js";
import {Errors} from "../errors.js";
import {KindError} from "../kind-error.js";

describe("KindError", () => {
    describe("#constructor", () => {
        it("should set the message", () => {
            // Arrange
            const message = "THE MESSAGE";

            // Act
            const error = new KindError(message);

            // Assert
            expect(error.message).toBe(message);
        });

        it("should set originalMessage to the message", () => {
            // Arrange
            const message = "THE MESSAGE";

            // Act
            const error = new KindError(message);

            // Assert
            expect(error.originalMessage).toBe(message);
        });

        it("should default kind", () => {
            // Arrange

            // Act
            const error = new KindError("MESSAGE");

            // Assert
            expect(error.kind).toBe(Errors.Unknown);
        });

        it("should set the kind", () => {
            // Arrange

            // Act
            const error = new KindError("MESSAGE", "CUSTOM_KIND");

            // Assert
            expect(error.kind).toBe("CUSTOM_KIND");
        });

        it("should default name", () => {
            // Arrange

            // Act
            const error = new KindError("MESSAGE");

            // Assert
            expect(error.name).toEndWith("Error");
        });

        it("should set the name", () => {
            // Arrange

            // Act
            const error = new KindError("MESSAGE", Errors.Unknown, {
                name: "CUSTOM_NAME",
            });

            // Assert
            expect(error.name).toEndWith("CUSTOM_NAMEError");
        });

        it.each(["N A M E", "NA\nME"])(
            "should throw if the name has whitespace like %s",
            (name) => {
                // Arrange

                // Act
                const act = () =>
                    new KindError("MESSAGE", Errors.Unknown, {name});

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            },
        );

        it.each(["K I N D", "KI\nND"])(
            "should throw if the kind has whitespace like %s",
            (kind) => {
                // Arrange

                // Act
                const act = () => new KindError("MESSAGE", kind);

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            },
        );

        it.each(["P R E F I X", "PRE\nFIX"])(
            "should throw if the prefix has whitespace like %s",
            (prefix) => {
                // Arrange

                // Act
                const act = () =>
                    new KindError("MESSAGE", "CUSTOM_KIND", {prefix});

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            },
        );

        it("should throw if stripStackFrames is < 0", () => {
            // Arrange

            // Act
            const act = () =>
                new KindError("MESSAGE", "CUSTOM_KIND", {stripStackFrames: -1});

            // Assert
            expect(act).toThrowErrorMatchingInlineSnapshot(
                `"stripStackFrames must be >= 0"`,
            );
        });

        it("should throw if minimumFrameCount is < 0", () => {
            // Arrange

            // Act
            const act = () =>
                new KindError("MESSAGE", "CUSTOM_KIND", {
                    minimumFrameCount: -1,
                });

            // Assert
            expect(act).toThrowErrorMatchingInlineSnapshot(
                `"minimumFrameCount must be >= 0"`,
            );
        });

        it("should clone metadata", () => {
            // Arrange
            const metadata = {
                foo: "bar",
            };
            const cloneMetadataSpy = jest
                .spyOn(CloneMetadata, "cloneMetadata")
                .mockReturnValue("CLONED_METADATA");

            // Act
            const error = new KindError("MESSAGE", Errors.Unknown, {metadata});

            // Assert
            expect(cloneMetadataSpy).toHaveBeenCalledWith(metadata);
            expect(error.metadata).toBe("CLONED_METADATA");
        });

        it("should get the normalized error info for the error being constructed, stripping frames", () => {
            // Arrange
            const normalizeSpy = jest.spyOn(ErrorInfo, "normalize");

            // Act
            const error = new KindError("MESSAGE", Errors.Unknown, {
                stripStackFrames: 1,
                minimumFrameCount: 2,
            });

            // Assert
            expect(normalizeSpy).toHaveBeenCalledWith(error, 1, 2);
        });

        it("should set the stack to the normalized value", () => {
            // Arrange
            const normalizedErrorInfo = new ErrorInfo(
                "NORMALIZED_NAME",
                "NORMALIZED_MESSAGE",
                ["normalizedstack1", "normalizedstack2"],
            );
            jest.spyOn(ErrorInfo, "normalize").mockReturnValue(
                normalizedErrorInfo,
            );

            // Act
            const error = new KindError("MESSAGE", Errors.Unknown);

            // Assert
            expect(error.stack).toBe(normalizedErrorInfo.standardizedStack);
        });

        it("should not set the message to the normalized value", () => {
            // Arrange
            const normalizedErrorInfo = new ErrorInfo(
                "NORMALIZED_NAME",
                "NORMALIZED_MESSAGE",
                ["normalizedstack1", "normalizedstack2"],
            );
            jest.spyOn(ErrorInfo, "normalize").mockReturnValue(
                normalizedErrorInfo,
            );

            // Act
            const error = new KindError("MESSAGE", Errors.Unknown);

            // Assert
            expect(error.message).toBe("MESSAGE");
        });

        describe("when cause is non-null", () => {
            it("should throw if cause is  not an Error", () => {
                // Arrange

                // Act
                const act = () =>
                    new KindError("MESSAGE", Errors.Unknown, {
                        // $FlowIgnore[incompatible-call]
                        cause: "NOT_AN_ERROR",
                    });

                // Assert
                expect(act).toThrowErrorMatchingInlineSnapshot(
                    `"cause must be an instance of Error"`,
                );
            });

            it("should get error info for the cause error", () => {
                // Arrange
                const cause = new Error("CAUSE_MESSAGE");
                const fromSpy = jest.spyOn(ErrorInfo, "from");

                // Act
                const act = () =>
                    new KindError("MESSAGE", Errors.Unknown, {
                        cause,
                        stripStackFrames: 1,
                        minimumFrameCount: 2,
                    });
                act();

                // Assert
                expect(fromSpy).toHaveBeenCalledWith(cause);
            });

            it("should combine the normalized error information of the constructed error and the regular error information from the causal error", () => {
                // Arrange
                const consequentialErrorInfo = new ErrorInfo(
                    "CONSEQUENCE_NAME",
                    "CONSEQUENCE_MESSAGE",
                    ["consequence1", "consequence2"],
                );
                const causalErrorInfo = new ErrorInfo(
                    "CAUSE_NAME",
                    "CAUSE_MESSAGE",
                    ["cause1", "cause2"],
                );
                const cause = new Error("CAUSE_MESSAGE");
                jest.spyOn(ErrorInfo, "normalize").mockReturnValue(
                    consequentialErrorInfo,
                );
                jest.spyOn(ErrorInfo, "from").mockReturnValue(causalErrorInfo);
                const fromConsequenceAndCauseSpy = jest.spyOn(
                    ErrorInfo,
                    "fromConsequenceAndCause",
                );

                // Act
                const act = () =>
                    new KindError("MESSAGE", Errors.Unknown, {cause});
                act();

                // Assert
                expect(fromConsequenceAndCauseSpy).toHaveBeenCalledWith(
                    consequentialErrorInfo,
                    causalErrorInfo,
                );
            });

            it("should set the message to the combined error info message", () => {
                // Arrange
                const cause = new Error("CAUSE_MESSAGE");
                const combinedErrorInfo = new ErrorInfo(
                    "COMBINED_NAME",
                    "COMBINED_MESSAGE",
                    ["combinedstack1", "combinedstack2"],
                );
                jest.spyOn(
                    ErrorInfo,
                    "fromConsequenceAndCause",
                ).mockReturnValue(combinedErrorInfo);

                // Act
                const error = new KindError("MESSAGE", Errors.Unknown, {cause});

                // Assert
                expect(error.message).toBe(combinedErrorInfo.message);
            });

            it("should have a combined message that cascades all causes", () => {
                // Arrange
                const cause1 = new Error("CAUSE_MESSAGE");
                const cause2 = new KindError(
                    "CAUSE_2_MESSAGE",
                    Errors.Unknown,
                    {cause: cause1},
                );

                // Act
                const result = new KindError("ROOT_MESSAGE", Errors.Unknown, {
                    cause: cause2,
                });

                // Assert
                expect(result.message).toMatchInlineSnapshot(`
                    "ROOT_MESSAGE
                    	caused by
                    		UnknownError: CAUSE_2_MESSAGE
                    	caused by
                    		Error: CAUSE_MESSAGE"
                `);
            });

            describe("when compositeStack is true", () => {
                it("should set the stack to the combined error standardized stack", () => {
                    // Arrange
                    const cause = new Error("CAUSE_MESSAGE");
                    const combinedErrorInfo = new ErrorInfo(
                        "COMBINED_NAME",
                        "COMBINED_MESSAGE",
                        ["combinedstack1", "combinedstack2"],
                    );
                    jest.spyOn(
                        ErrorInfo,
                        "fromConsequenceAndCause",
                    ).mockReturnValue(combinedErrorInfo);

                    // Act
                    const error = new KindError("MESSAGE", Errors.Unknown, {
                        cause,
                        compositeStack: true,
                    });

                    // Assert
                    expect(error.stack).toBe(
                        combinedErrorInfo.standardizedStack,
                    );
                });
            });

            describe.each([undefined, null, false])(
                "when composite stack is %s",
                (compositeStackValue) => {
                    it("should set the stack to the normalized stack, not the combined stack", () => {
                        // Arrange
                        const normalizedErrorInfo = new ErrorInfo(
                            "NORMALIZED_NAME",
                            "NORMALIZED_MESSAGE",
                            ["normalizedstack1", "normalizedstack2"],
                        );
                        jest.spyOn(ErrorInfo, "normalize").mockReturnValue(
                            normalizedErrorInfo,
                        );
                        const cause = new Error("CAUSE_MESSAGE");
                        const combinedErrorInfo = new ErrorInfo(
                            "COMBINED_NAME",
                            "COMBINED_MESSAGE",
                            ["combinedstack1", "combinedstack2"],
                        );
                        jest.spyOn(
                            ErrorInfo,
                            "fromConsequenceAndCause",
                        ).mockReturnValue(combinedErrorInfo);

                        // Act
                        const error = new KindError("MESSAGE", Errors.Unknown, {
                            cause,
                            compositeStack: compositeStackValue,
                        });

                        // Assert
                        expect(error.stack).toBe(
                            normalizedErrorInfo.standardizedStack,
                        );
                    });
                },
            );
        });
    });
});
