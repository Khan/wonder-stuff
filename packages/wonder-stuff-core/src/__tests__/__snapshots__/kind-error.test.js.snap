// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KindError #constructor should throw if the kind has whitespace like K I N D 1`] = `"kind must not contain whitespace"`;

exports[`KindError #constructor should throw if the kind has whitespace like KI
ND 1`] = `"kind must not contain whitespace"`;

exports[`KindError #constructor should throw if the name has whitespace like N A M E 1`] = `"name must not contain whitespace"`;

exports[`KindError #constructor should throw if the name has whitespace like NA
ME 1`] = `"name must not contain whitespace"`;

exports[`KindError #constructor should throw if the prefix has whitespace like P R E F I X 1`] = `"prefix must not contain whitespace"`;

exports[`KindError #constructor should throw if the prefix has whitespace like PRE
FIX 1`] = `"prefix must not contain whitespace"`;
